### ISSUE TITLE
How to visualize the predicted results of the model

### ISSUE THREAD

**Issue by xu758** [InhwanBae/ETH-UCY-Trajectory-Visualizer#1]
hello，when I reproduce the visualization code, it only generates the ground truth trajectories. I would like to know how to generate the predicted trajectories from the model.

**Comment by InhwanBae** [comment]
Hi @xu758,

You can use two approaches to visualize the trajectory prediction results of your own model. The first is to integrate my code into your evaluation pipeline, and the second is to bundle the prediction results into a list and pickle it for loading in the visualizer. I recommend the second approach, and you can use my commented lines of code in the `visualize_ethucy.py` file.

**Comment by liutongli** [comment]
Hello, I encountered a problem during visualization. How can I bundle the prediction results into a list?

**Comment by KimNaser** [comment]
@liutongli I gave up, did solve the issue?
@InhwanBae thank you your work. Really, it is imazing. could you add a code for that, please.? I would like to use your model for this [paper](https://github.com/InhwanBae/GPGraph).

**Comment by InhwanBae** [comment]
Hi @liutongli and @KimNaser,

To summarize, you just need to use the predicted trajectory array to plot the temporal xy-coordinate sequence for each pedestrians. To bundle the predicted results for [**GPGraph**](https://github.com/InhwanBae/GPGraph/blob/main/test.py), you can use the following simple and straightforward approach:

```python
# test.py
line 11: def test(SAMPLES=20, TRIALS=1):
line 13: all_pred_traj = []
line 52: 
best_traj_idx = (V_absl-V_pred_traj_gt).norm(p=2, dim=-1)[:, -1].argmin(dim=0)  # Get the best traj based on FDE
all_pred_traj.append(V_absl[best_traj_idx].permute(1, 0, 2).detach().cpu().numpy())  # shape = [pedid, time, coordinate]
line 67:
import pickle
with open(file='GPGraph-ETH.pkl', mode='wb') as f:
    pickle.dump(all_pred_traj, f)
```

If you’re struggling with any other part, feel free to let me know.

**Comment by zhangyu2234** [comment]
> Hi @liutongli and @KimNaser,
> 
> To summarize, you just need to use the predicted trajectory array to plot the temporal xy-coordinate sequence for each pedestrians. To bundle the predicted results for [**GPGraph**](https://github.com/InhwanBae/GPGraph/blob/main/test.py), you can use the following simple and straightforward approach:
> 
> ```python
> # test.py
> line 11: def test(SAMPLES=20, TRIALS=1):
> line 13: all_pred_traj = []
> line 52: 
> best_traj_idx = (V_absl-V_pred_traj_gt).norm(p=2, dim=-1)[:, -1].argmin(dim=0)  # Get the best traj based on FDE
> all_pred_traj.append(V_absl[best_traj_idx].permute(1, 0, 2).detach().cpu().numpy())  # shape = [pedid, time, coordinate]
> line 67:
> import pickle
> with open(file='GPGraph-ETH.pkl', mode='wb') as f:
>     pickle.dump(all_pred_traj, f)
> ```
> 
> If you’re struggling with any other part, feel free to let me know.

Hello, when I execute all_pred_traj.append(V_absl[best_traj_idx].permute(1, 0, 2).detach().cpu().numpy()), an error message appears, saying number of dimensions in the tensor input does not match the length of the desired ordering of dimensions i.e. input.dim() = 4 is not equal to len(dims) = 3. This is because V_abs is a four-dimensional tensor. How should I solve this problem?

**Comment by InhwanBae** [comment]
Hi @zhangyu2234,

I’ve updated the code at line 52, and it should work correctly now. For reference, this line functions the same as the one shown below.

```python
# Short code
all_pred_traj.append(V_absl.permute(0, 2, 1, 3)[best_traj_idx, torch.arange(len(best_traj_idx))].detach().cpu().numpy())  # shape = [pedid, time, coordinate]

# Long code
temp = np.zeros([len(best_traj_idx), V_absl.size(1), V_absl.size(3)])  # shape = [pedid, time, coordinate]
for ped_id in range(len(best_traj_idx)):
    temp[ped_id] = V_absl[:, :, ped_id][best_traj_idx[ped_id]].detach().cpu().numpy()
all_pred_traj.append(temp)
```

**Comment by zhangyu2234** [comment]
> ```python
> import pickle
> with open(file='GPGraph-ETH.pkl', mode='wb') as f:
>     pickle.dump(all_pred_traj, f)
> ```
Thank you very much for your reply. I have another question. When executing this line of code, traj = dataset.image_to_world(data[idx][n]) + 0.5, it will prompt TrajectoryDataset' object has no attribute 'image_to_world. Could you give me the content of image_to_world? Thank you very much for your work!

**Comment by zhangyu2234** [comment]
> 你好@zhangyu2234，
> 
> 我更新了第 52 行的代码，现在它应该可以正常工作了。作为参考，此行的功能与下面显示的相同。
> 
> ```python
> # Short code
> all_pred_traj.append(V_absl.permute(0, 2, 1, 3)[best_traj_idx, torch.arange(len(best_traj_idx))].detach().cpu().numpy())  # shape = [pedid, time, coordinate]
> 
> # Long code
> temp = np.zeros([len(best_traj_idx), V_absl.size(1), V_absl.size(3)])  # shape = [pedid, time, coordinate]
> for ped_id in range(len(best_traj_idx)):
>     temp[ped_id] = V_absl[:, :, ped_id][best_traj_idx[ped_id]].detach().cpu().numpy()
> all_pred_traj.append(temp)
> ```

Thank you very much for your reply. I have another question. When executing this line of code, traj = dataset.image_to_world(data[idx][n]) + 0.5, it will prompt TrajectoryDataset' object has no attribute 'image_to_world. Could you give me the content of image_to_world? Thank you very much for your work!

**Comment by InhwanBae** [comment]
Hi @zhangyu2234, Transformation codes are written in `utils.homography`. You can update it as follows.

```python
line 7: from utils.homography import world2image, image2world
line 50: traj = image2world(data[idx][n], H=dataset.h_mat) + 0.5
```

**Comment by zhangyu2234** [comment]
Dear Author,

I hope this email finds you well.

First of all, I would like to express my sincere gratitude for your unwavering support in addressing my questions. Your patience and willingness to help have been invaluable and have greatly aided my progress in the project.

After running the trajectory prediction code, I noticed that the predicted trajectories for the ETH dataset, as shown in the attached image, are clearly incorrect. The trajectories deviate significantly from what I expected. At this point, I have been unable to identify the root cause of the issue. Could it possibly be related to some aspect of data preprocessing, or might there be a minor issue in the code that I have overlooked?

I would greatly appreciate it if you could assist me once again in pinpointing the issue and offer any advice on how to resolve it. Thank you again for your time and continued support.

I look forward to your reply, and I wish you all the best.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "InhwanBae/ETH-UCY-Trajectory-Visualizer"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年1月5日(星期天) 下午4:54
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [InhwanBae/ETH-UCY-Trajectory-Visualizer] How to visualize the predicted results of the model (Issue #1)





 
Hi @zhangyu2234, Transformation codes are written in utils.homography. You can update it as follows.
 line 7: from utils.homography import world2image, image2world line 50: traj = image2world(data[idx][n], H=dataset.h_mat) + 0.5

**Comment by zhangyu2234** [comment]
> Hi @zhangyu2234, Transformation codes are written in `utils.homography`. You can update it as follows.
> 
> ```python
> line 7: from utils.homography import world2image, image2world
> line 50: traj = image2world(data[idx][n], H=dataset.h_mat) + 0.5
> ```
Dear Author,

I hope this email finds you well.

First of all, I would like to express my sincere gratitude for your unwavering support in addressing my questions. Your patience and willingness to help have been invaluable and have greatly aided my progress in the project.

After running the trajectory prediction code, I noticed that the predicted trajectories for the ETH dataset, as shown in the attached image, are clearly incorrect. The trajectories deviate significantly from what I expected. At this point, I have been unable to identify the root cause of the issue. Could it possibly be related to some aspect of data preprocessing, or might there be a minor issue in the code that I have overlooked?

I would greatly appreciate it if you could assist me once again in pinpointing the issue and offer any advice on how to resolve it. Thank you again for your time and continued support.

I look forward to your reply, and I wish you all the best.

![Uploading traj_34_9750.png…]()

**Comment by zhangyu2234** [comment]
> Hi @zhangyu2234, Transformation codes are written in `utils.homography`. You can update it as follows.
> 
> ```python
> line 7: from utils.homography import world2image, image2world
> line 50: traj = image2world(data[idx][n], H=dataset.h_mat) + 0.5
> ```

Dear Author,

I hope this email finds you well.

First of all, I would like to express my sincere gratitude for your unwavering support in addressing my questions. Your patience and willingness to help have been invaluable and have greatly aided my progress in the project.

After running the trajectory prediction code, I noticed that the predicted trajectories for the ETH dataset, as shown in the attached image, are clearly incorrect. The trajectories deviate significantly from what I expected. At this point, I have been unable to identify the root cause of the issue. Could it possibly be related to some aspect of data preprocessing, or might there be a minor issue in the code that I have overlooked?

I would greatly appreciate it if you could assist me once again in pinpointing the issue and offer any advice on how to resolve it. Thank you again for your time and continued support.

I look forward to your reply, and I wish you all the best.
![Uploading traj_31_9720.png…]()

**Comment by zhangyu2234** [comment]
> Hi @zhangyu2234, Transformation codes are written in `utils.homography`. You can update it as follows.
> 
> ```python
> line 7: from utils.homography import world2image, image2world
> line 50: traj = image2world(data[idx][n], H=dataset.h_mat) + 0.5
> ```

Dear Author,

I hope this email finds you well.

First of all, I would like to express my sincere gratitude for your unwavering support in addressing my questions. Your patience and willingness to help have been invaluable and have greatly aided my progress in the project.

After running the trajectory prediction code, I noticed that the predicted trajectories for the ETH dataset, as shown in the attached image, are clearly incorrect. The trajectories deviate significantly from what I expected. At this point, I have been unable to identify the root cause of the issue. Could it possibly be related to some aspect of data preprocessing, or might there be a minor issue in the code that I have overlooked?

I would greatly appreciate it if you could assist me once again in pinpointing the issue and offer any advice on how to resolve it. Thank you again for your time and continued support.

I look forward to your reply, and I wish you all the best.
![Uploading traj_31_9720.png…](ETH)

**Comment by zhangyu2234** [comment]
Dear Author,

I hope this email finds you well.

First of all, I would like to express my sincere gratitude for your unwavering support in addressing my questions. Your patience and willingness to help have been invaluable and have greatly aided my progress in the project.

After running the trajectory prediction code, I noticed that the predicted trajectories for the ETH dataset, as shown in the attached image, are clearly incorrect. The trajectories deviate significantly from what I expected. At this point, I have been unable to identify the root cause of the issue. Could it possibly be related to some aspect of data preprocessing, or might there be a minor issue in the code that I have overlooked?
I would greatly appreciate it if you could assist me once again in pinpointing the issue and offer any advice on how to resolve it. Thank you again for your time and continued support.

I look forward to your reply, and I wish you all the best.

![traj_31_9720](https://github.com/user-attachments/assets/931cdc90-8b7e-4e20-b996-32fef39ec5bd)

**Comment by Whitney-9** [comment]
Hello, how to visualize the SingularTrajectory's prediction result?

**Comment by InhwanBae** [comment]
Hi @Whitney-9 

For SingularTrajectory, you can use the following code.

```python
line 327: all_pred_traj = []
line 335:
best_traj_idx = (output["recon_traj"]-pred_traj).norm(p=2, dim=-1)[:, :, -1].argmin(dim=0)  # Get the best traj based on FDE
temp = np.zeros([len(best_traj_idx), output["recon_traj"].size(2), output["recon_traj"].size(3)])  # shape = [pedid, time, coordinate]
for ped_id in range(len(best_traj_idx)):
    temp[ped_id] = output["recon_traj"][:, ped_id][best_traj_idx[ped_id]].detach().cpu().numpy()
all_pred_traj.append(temp)
line 339:
import pickle
with open(file='SingularTrajectory-ETH.pkl', mode='wb') as f:
    pickle.dump(all_pred_traj, f)

````

**Comment by Joe-zy01** [comment]
Hello, I am very interested in your work. How to visualize the Graph-TERN's prediction result? Could you share more about it?
